nova/virt/vmwareapi/vmops.py

```diff
@@ -2772,7 +2913,133 @@ def live_migration(self, instance, migrate_data, volume_mapping):
             config_spec.device = device
             device_config_spec.append(config_spec)
 
-        vm_util.relocate_vm(self._session, vm_ref, spec=relocate_spec)
+        try:
+            vm_util.relocate_vm(self._session, vm_ref, spec=relocate_spec)
+        except vexc.VimException as e:
+            target.delete_config_drive_files(context, instance, new_cdroms)
+            raise e
+
+        self.delete_config_drive_files(context, instance, original_cdroms)
+
+        if new_cdroms:
+            # Technically a post-live-migration task, but we have no
+            # way of passing those changes without API changes to the
+            # destination compute-host
+            try:
+                target.reconfigure_vm_device_change(context, instance,
+                                                    new_cdroms)
+            except messaging.RemoteError:
+                # Failing now would stop the instance reflecting
+                # the new host
+                LOG.exception("Failed to reconfiguring cdroms. "
+                              "Swallowing error as we can't go back now",
+                               instance=instance)
```

Comment:
But also nobody will ever notice we lost the cdrom unless somebody checks sentry by accident and feels responsible to fix this ...

Can we add an action-event for that part? I know there can be multiple events for a server action, but I don't know if we're able to add them here, but that way it would be possible to record such non-fatal errors and the calling operator could just check via API if everything succeeded.

Reply:
> But also nobody will ever notice we lost the cd-rom unless somebody checks sentry by accident and feels responsible to fix this ...

Well, we do not lose it. It will "just" not be connected. And if no one notices, then we are fine :).

The reason why it is written that way lies in the fact, that if we raise an exception here,
then Nova will record the VM as being still on the old host, despite the actual migration having succeeded.

I am not too much a fan of logging to another additional location.

I see following ways to improve it:
- I can write a hammer/screw script to check and fix that particular error.
- I can search for a way to transport that information to the post-migration step in the destination node and error out there
  - Either via an API change
  - Backchannel: Instead of configuring it right away, I store the value (in memory) on the destination node and then apply the configuration change in the post-migration step. Reminder: The whole migration task is in memory of the conductor node, so it isn't restart safe either (yet)

Reply:
I just don't see much value in raising an exception-level log if nobody will take care of that and if it's important for the operator doing the migration. instance-action-events would be more visible and could be checked easier.

hammer/screw is fine and all, but someone will have to call it. if we can check on the migration task that some optional operation didn't succeed, the operator could run the screw command right away and have a look.

so I think this is more about visibility for me than about "solving it right", as I would think the visibility could come at much lower costs.

Reply:
> hammer/screw is fine and all, but someone will have to call it. if we can check on the migration task that some optional operation didn't succeed, the operator could run the screw command right away and have a look.

My idea is, that the operator migrates via `screw os migrate-vms`, which could do that check automatically. I would simply call the vcenter an check if the cdrom is properly attached as part of the whole post-live-migration check already in-place, which reports then the error to the user.

I am reluctant to log things to the database, as I consider it bad style... but it seems that ship has sailed, as it is implemented quite conveniently:
That change should do the trick:
```python
@compute_utils.wrap_instance_event(prefix="vmwareapi")
def reconfigure_vm_device_change(self, context, instance, devices):
    ....
```

Reply:
The implied question: Would it be sufficient to report the error to the operator via screw, would you still rather the event being logged to the database?

Reply:
Sure, I can do it. I just want to clarify it more, because I still don't seem to get it, why that is preferable, and I may be missing something. So it is more for my own understanding.

I believe the only error with enough visibility and acted upon would be one in the VM state, which I consider a reasonable ask. I am only hesitant about that because the solutions are not "nice".

I agree, that any other error, such as the log message, or sentry report requires a second step, and is therefor limited in their visibility. As I understand though, that also applies to the instance-actions and their events. That is what I am thinking I am missing.

I think, one has to call `GET /servers/{server_id}/os-instance-actions/{request_id}` to get the individual events of an action. 
I am clearly biased against that interface for stylistic reasons, and was wondering, if it is a common thing people do.
But as far as I can see in the logs from the three larges regions, it happened only once in the last two weeks with your credentials.

On a side-note: Apparently a go-httpclient (I assume kubernetes) is polling the `os-instance-actions` with `changes-since`.

Reply:
According to Nova's docs, clients are supposed to query the `os-instance-actions` to retrieve the state and result of asynchronous tasks. I know that people don't really do that, but I don't know why. I try to educate them every time this comes up.

Sentry messages go away after a short time and also contain just a sample of the events. Therefore, they seem unfit if there are multiple migrations in a day.

Log messages from screw basically go away immediately and the nova-compute logs after a couple of weeks max.

The instance-actions stay around, though, so we can still debug and verify a problem after some time when the customer tries to access the configdrive again, because they rebooted in some maintenance window long time after the migration. Therefore, I like this place to store errors.